lib/NativeCall.rakumod

```diff
@@ -260,6 +263,144 @@ sub resolve-libname($libname) {
     $libname.platform-library-name.Str
 }
 
+use MONKEY-SEE-NO-EVAL;
+my $use-dispatcher = so $*RAKU.compiler.?supports-op('dispatch_v') && EVAL q:to/CODE/;
+    use nqp;
+    sub raku-native-dispatch-deproxy($callee, *@args) {
+        $callee(|@args) # implicitly fetches values from proxies
+    }
+    my &raku-nativecall := -> $capture is raw {
+        my $track_callee := nqp::dispatch('boot-syscall', 'dispatcher-track-arg', $capture, 0);
+        nqp::dispatch('boot-syscall', 'dispatcher-guard-literal', $track_callee);
+        my $callee := nqp::captureposarg($capture, 0);
+        $callee.setup;
+
+        my Mu $args := nqp::dispatch('boot-syscall', 'dispatcher-drop-arg', $capture, 0);
+        my $signature := nqp::getattr($callee, Code, '$!signature');
+        my int $readonly = nqp::getattr_i($signature, Signature, '$!readonly');
+        my int $pos-args = nqp::captureposelems($args);
+        my int $i = 0;
+        my int $continue = 1;
+        while $i < $pos-args and $continue {
+            # If it should be passed read only, and it's an object...
+            if nqp::captureposprimspec($args, $i) == 0 {
+                # If it's in a Scalar container...
+                my $track-arg := nqp::dispatch('boot-syscall', 'dispatcher-track-arg', $args, nqp::unbox_i($i));
+                nqp::dispatch('boot-syscall', 'dispatcher-guard-type', $track-arg);
+                nqp::dispatch('boot-syscall', 'dispatcher-guard-concreteness', $track-arg);
+                my $arg := nqp::captureposarg($args, $i);
+                my $track-value;
+                my $cstr = False;
+                if nqp::isconcrete_nd($arg) && nqp::what_nd($arg) =:= Scalar {
```

Comment:
I wonder it we should permit subclasses of `Scalar` too (make it `nqp::istype_nd($arg, Scalar)`).

Reply:
If we permit subclasses here, we should probably permit them in pass-decontainerized (where I stole this from) as well: https://github.com/rakudo/rakudo/blob/master/src/vm/moar/dispatchers.nqp#L2873

Reply:
Fair. Though in the place you stole it from it's an optimization issue (worst case is the callee does the deconting), here I think it's a correctness one (there's nothing else to do the deconting).